elements/lisk-p2p/src/directory/new_peers.ts

```diff
@@ -0,0 +1,277 @@
+	}
+
+	public getBucketId(ipAddress: string): number {
+		return getBucket({
+			secret: this._secret,
+			peerType: PEER_TYPE.NEW_PEER,
+			targetAddress: ipAddress,
+		});
+	}
+
+	public findPeer(peerInfo: P2PPeerInfo): boolean {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		if (bucket && bucket.get(constructPeerIdFromPeerInfo(peerInfo))) {
+			return true;
+		}
+
+		return false;
+	}
+
+	public updatePeer(peerInfo: P2PPeerInfo): boolean {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		if (bucket) {
+			const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+			const foundPeer = bucket.get(incomingPeerId);
+			if (foundPeer) {
+				const updatedNewPeerInfo: NewPeerInfo = {
+					peerInfo: { ...foundPeer.peerInfo, ...peerInfo },
+					dateAdded: foundPeer.dateAdded,
+				};
+
+				bucket.set(incomingPeerId, updatedNewPeerInfo);
+				this._newPeerMap.set(bucketId, bucket);
+
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public removePeer(peerInfo: P2PPeerInfo): boolean {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+		if (bucket && bucket.get(incomingPeerId)) {
+			const success = bucket.delete(incomingPeerId);
+			this._newPeerMap.set(bucketId, bucket);
+
+			return success;
+		}
+
+		return false;
+	}
+
+	public getPeer(peerInfo: P2PPeerInfo): P2PPeerInfo | undefined {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+		if (bucket) {
+			const newPeer = bucket.get(incomingPeerId);
+
+			return newPeer ? newPeer.peerInfo : undefined;
+		}
+
+		return undefined;
+	}
+
+	public addPeer(peerInfo: P2PPeerInfo): AddPeerOutcome {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+
+		if (bucket && bucket.get(incomingPeerId)) {
+			return {
+				success: false,
+				isEvicted: false,
+			};
+		}
+
+		if (bucket) {
+			const newPeerInfo = {
+				peerInfo,
+				numOfConnectionFailures: 0,
+				dateAdded: new Date(),
+			};
+
+			if (bucket.size < this._newPeerBucketSize) {
+				bucket.set(incomingPeerId, newPeerInfo);
+				this._newPeerMap.set(bucketId, bucket);
+
+				return {
+					success: true,
+					isEvicted: false,
+				};
+			} else {
+				const evictedPeer = this._evictPeer(bucketId);
+				bucket.set(incomingPeerId, newPeerInfo);
+				this._newPeerMap.set(bucketId, bucket);
+
+				return {
+					success: true,
+					isEvicted: true,
+					evictedPeer: evictedPeer.peerInfo,
+				};
+			}
+		}
+
+		return {
+			success: false,
+			isEvicted: false,
+		};
+	}
+
+	public failedConnectionAction(incomingPeerInfo: P2PPeerInfo): boolean {
```

Comment:
In newPeers since we are only removing a peer after a disconnection we are doing what remove function does, but for consistency with `triedPeers` we want to keep such a function in newPeers too and also in future if we have to change it then we can simply implement it (for example, deletion after 2 disconnections in newPeers)